### PR TITLE
[RO-269] UUID identifier generation strategy for Line items

### PR DESCRIPTION
**Pull request by Skyld**
Ticket: https://oat-sa.atlassian.net/browse/RO-269

Second PR of switching from sequential id generation strategy to uuid.

### Changed
- Changed identifier generation strategy of entities from `auto increment` to `uuid`.

### CODE REVIEW

File: src/Command/Cache/LineItemCacheWarmerCommand.php [hunk 1]
```diff
@@ -117,14 +117,14 @@ protected function execute(InputInterface $input, OutputInterface $output): int
...
-                $id = (int)$lineItem->getId();
+                $id = $lineItem->getId();
                 $lineItemCacheKey = $this->lineItemCacheIdGenerator->generate($id);
 
                 $this->resultCacheImplementation->delete($lineItemCacheKey);
                 $this->lineItemRepository->findOneById($id);
 
                 $this->logger->info(
-                    sprintf('Result cache for Line Item Id %d have been successfully warmed up.', $id),
+                    sprintf("Result cache for Line Item with id = '%s' have been successfully warmed up.", (string)$id),
```
Comment (luisfn) [inline src/Command/Cache/LineItemCacheWarmerCommand.php:127]: Not sure what suites best but is also possible to use $id->jsonSerialize() instead of casting.

File: src/Command/Ingester/LineItemIngesterCommand.php [hunk 2]
```diff
@@ -172,6 +173,7 @@ private function createLineItem(array $rawLineItem): LineItem
     {
         $lineItem =
             (new LineItem())
+                ->setId(new UuidV6())
```
Comment (luisfn) [inline src/Command/Ingester/LineItemIngesterCommand.php:176]: Any idea of performance impact?

File: src/Command/ModifyEntity/LineItem/LineItemChangeDatesCommand.php [hunk 3]
```diff
@@ -200,8 +201,8 @@ protected function execute(InputInterface $input, OutputInterface $output): int
 
                 $this->logger->info(
                     sprintf(
-                        'New dates were set for line item with: "%d"',
-                        $lineItem->getId()
+                        'New dates were set for line item with: "%s"',
+                        (string)$lineItem->getId()
```
Comment (luisfn) [inline src/Command/ModifyEntity/LineItem/LineItemChangeDatesCommand.php:205]: same as previously, not sure what is best, casting or use jsonSerialize()

File: src/Command/ModifyEntity/LineItem/LineItemChangeDatesCommand.php [hunk 4]
```diff
@@ -237,20 +238,22 @@ protected function execute(InputInterface $input, OutputInterface $output): int
...
     private function initializeLineItemIdsOption(InputInterface $input): void
     {
-        $lineItemIds = array_filter(
-            explode(',', (string)$input->getOption(self::OPTION_LINE_ITEM_IDS)),
-            static function (string $value): bool {
-                return !empty($value) && (int)$value > 0;
-            }
-        );
+        try {
+            $this->lineItemIds = array_map(
```
Comment (luisfn) [inline src/Command/ModifyEntity/LineItem/LineItemChangeDatesCommand.php:242]: Remember to change help/docs where it give examples with numeric IDs

**Review by luisfn — COMMENTED** [review]
Another nice PR. Seems good to me and just added some notes

File: src/Command/Ingester/LineItemIngesterCommand.php (hunk 2 above)
Comment (andreluizmachado) [reply src/Command/Ingester/LineItemIngesterCommand.php:176]: Do we have to check The increase of the database (Index, data)? Does it imply more costs?

File: src/Command/ModifyEntity/LineItem/LineItemChangeStateCommand.php [hunk 5]
```diff
@@ -171,11 +181,10 @@ protected function execute(InputInterface $input, OutputInterface $output): int
...
                 $this->logger->info(
                     sprintf(
-                        'The operation: "%s" was executed for Line Item with id: "%d"',
+                        "The operation: '%s' was executed for Line Item with id: '%s'",
                         $toggle,
-                        $lineItem->getId()
-                    ),
-                    ['slug' => $lineItem->getSlug(), 'uri' => $lineItem->getUri()]
+                        (string)$lineItem->getId()
+                    )
```
Comment (andreluizmachado) [inline src/Command/ModifyEntity/LineItem/LineItemChangeStateCommand.php:187]: Why did you remove the context?

File: src/EventListener/Doctrine/LineItemLoaderListener.php [hunk 6]
```diff
@@ -43,7 +43,7 @@ public function postLoad(Assignment $assignment): void
     {
         //we're forcing the doctrine to load the line item by using cache instead of database query
         $assignment->setLineItem(
-            $this->lineItemRepository->findOneById($assignment->getLineItemId())
+            $this->lineItemRepository->findOneById($assignment->getLineItem()->getId())
```
Comment (andreluizmachado) [inline src/EventListener/Doctrine/LineItemLoaderListener.php:46]: I think that we have a cache issue here. This call will make a DB query I think. Can you check it, please?

**Review by andreluizmachado — CHANGES_REQUESTED** [review]
Nice Job @Skyld. I have just a few comments.

File: src/Command/Cache/LineItemCacheWarmerCommand.php (hunk 1 above)
Comment (uncleempty) [reply src/Command/Cache/LineItemCacheWarmerCommand.php:127]: Agree with Luis on possible usage, but IMO - doesn't really matter

File: src/Command/Ingester/LineItemIngesterCommand.php (hunk 2 above)
Comment (uncleempty) [reply src/Command/Ingester/LineItemIngesterCommand.php:176]: I have a more interesting question, not for this file, but for the whole PR itself:
what is the reason to go with UUIDv6, instead of ULID (though both are supported by 5.x branch of symfony as I see)?
Because UUIDv6 is not part of the standard, and status of RFC (?) is undefined (at least I didn't manage to find recent info on status).
So for me it's more a question of: should we really go with smth non-standard (though supported) or with smth "standard"?
p.s.
As I can see with my google-fu:
- https://symfony.com/doc/current/components/uid.html#ulids
- https://github.com/prooph/event-store/issues/361 (just the discussion around using ULIDs)
p.p.s
Though I didn't find it - I though that there already was a some kind of discussion either in slack or elsewhere on ULIDs limitations and what to use, maybe you can point me

File: src/Command/ModifyEntity/LineItem/LineItemChangeDatesCommand.php (hunk 4 above)
Comment (uncleempty) [reply src/Command/ModifyEntity/LineItem/LineItemChangeDatesCommand.php:242]: Also: IMO, a good point to raise a discussion - how _this_ should be released, as 2.smth or _3.smth_? Because for me, at least partially, those changes seem a bit breaking :upside_down_face:

File: src/Command/ModifyEntity/LineItem/LineItemChangeStateCommand.php (hunk 5 above)
Comment (uncleempty) [reply src/Command/ModifyEntity/LineItem/LineItemChangeStateCommand.php:187]: I dunno Dani's opinion on this, but as for me, _in this_ command we already have more than enough context. How do you think?

File: src/DataTransferObject/AssignmentDto.php [hunk 7]
```diff
@@ -36,7 +38,7 @@ class AssignmentDto
     /** @var int|null */
     private $userId;
 
-    public function __construct(string $state, int $lineItemId, string $username, int $userId = null)
+    public function __construct(string $state, UuidV6 $lineItemId, string $username, int $userId = null)
```
Comment (uncleempty) [inline src/DataTransferObject/AssignmentDto.php:41]: Not that it really matters or is in any standards, but:
IMO - Object arguments should go first, as it improved readability a bit

File: src/Repository/Criteria/FindLineItemCriteria.php [hunk 8]
```diff
@@ -22,15 +22,17 @@
...
 {
-    /** @var int[] */
+    /** @var UuidV6[] */
     private $lineItemIds = [];
 
     /** @var string[] */
     private $lineItemSlugs = [];
 
-    public function addLineItemIds(int ...$lineItemIds): self
+    public function addLineItemIds(UuidV6 ...$lineItemIds): self
```
Comment (uncleempty) [inline src/Repository/Criteria/FindLineItemCriteria.php:35]: Just a stupid question:
why the method is `add...` instead of `set...`, since we are not adding inside, but setting?

File: src/Repository/LineItemRepository.php [hunk 9]
```diff
@@ -61,18 +62,18 @@ public function findAllAsCollection(): LineItemCollection
...
      * @throws EntityNotFoundException
      */
-    public function findOneById(int $id): LineItem
+    public function findOneById(UuidV6 $id): LineItem
     {
         $lineItem = $this->createQueryBuilder('l')
             ->select('l')
             ->where('l.id = :id')
-            ->setParameter('id', $id)
+            ->setParameter('id', $id, 'uuid')
```
Comment (uncleempty) [inline src/Repository/LineItemRepository.php:70]: Can we have a way we are working with uuids in repos unified more or less?
Meaning that in line 70 we are using third arg to tell Doctrine the type,
but on line 93 we are converting to binary form
p.s.
I know that for line 93, and `... IN()...` it's easier to go with binary form, but why not using `toBinary` call in there?

**Review by uncleempty — CHANGES_REQUESTED** [review]
Good job Dani! :+1: 
p.s.
From my side it's mostly questions, not real change requests

File: src/Command/Ingester/LineItemIngesterCommand.php (hunk 2 above)
Comment (Skyld) [reply src/Command/Ingester/LineItemIngesterCommand.php:176]: I think we discussed this today on Slack. Can I close this conversation @uncleempty ?

File: src/Command/ModifyEntity/LineItem/LineItemChangeStateCommand.php (hunk 5 above)
Comment (Skyld) [reply src/Command/ModifyEntity/LineItem/LineItemChangeStateCommand.php:187]: Yes, that was the reason, line item id is enough in the message, the rest of the context was not relevant here (because we change only the status)

File: src/Command/ModifyEntity/LineItem/LineItemChangeDatesCommand.php (hunk 4 above)
Comment (Skyld) [reply src/Command/ModifyEntity/LineItem/LineItemChangeDatesCommand.php:243]: This is an interesting question, let's discuss it together with the guys, I have a proposal :)

File: src/EventListener/Doctrine/LineItemLoaderListener.php (hunk 6 above)
Comment (Skyld) [reply src/EventListener/Doctrine/LineItemLoaderListener.php:46]: I am going to profile it, good catch!

File: src/Command/Ingester/LineItemIngesterCommand.php (hunk 2 above)
Comment (Skyld) [reply src/Command/Ingester/LineItemIngesterCommand.php:176]: Resolving the conversation, if you think we need more discussion around this, let's reopen it.

File: src/Command/ModifyEntity/LineItem/LineItemChangeDatesCommand.php (hunk 4 above)
Comment (Skyld) [reply src/Command/ModifyEntity/LineItem/LineItemChangeDatesCommand.php:243]: Let's postpone this question until we merge back the integration branch back to develop

File: src/Repository/LineItemRepository.php (hunk 9 above)
Comment (Skyld) [reply src/Repository/LineItemRepository.php:70]: Checking

Comment (Skyld) [reply src/Repository/LineItemRepository.php:70]: Do you have any suggestion @uncleempty ?

**Review by andreluizmachado — APPROVED** [review]
- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)

File: src/Command/ModifyEntity/LineItem/LineItemChangeDatesCommand.php (hunk 4 above)
Comment (luisfn) [reply src/Command/ModifyEntity/LineItem/LineItemChangeDatesCommand.php:243]: I think you missed to update the command help (L94-95). There are some numeric ids there

And also the docs:

https://github.com/oat-sa/simple-roster/blob/feature/line-item-uuid/docs/cli/modify-entity-line-item-change-dates-command.md
https://github.com/oat-sa/simple-roster/blob/feature/line-item-uuid/docs/cli/modify-entity-line-item-change-state-command.md

**Review by luisfn — CHANGES_REQUESTED** [review]
On the issue with docs and we are good

File: src/Command/ModifyEntity/LineItem/LineItemChangeDatesCommand.php (hunk 4 above)
Comment (Skyld) [reply src/Command/ModifyEntity/LineItem/LineItemChangeDatesCommand.php:246]: Fixed

**Review by luisfn — APPROVED** [review]
**Review Checklist**
- [X] New code is covered by tests (if applicable)
- [X] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [X] New code is respecting code style rules
- [X] New code is respecting best practices
- [X] New code is not subject to concurrency issues (if applicable)
- [X] Feature is working correctly on my local machine (if applicable)
- [X] Acceptance criteria are respected
- [X] Pull request title and description are meaningful
- [ ] Changelog is updated according to changes (if applicable)
- [X] Documentation is updated according to changes (if applicable)
